### PR TITLE
fix(metadata): extract epub titles per spec with regards to ordering

### LINKED ISSUE #1636: EPUB metadata extractor does not fallback to order-based title parsing
### Discussed in https://github.com/orgs/grimmory-tools/discussions/627

<div type='discussions-op-text'>

<sup>Originally posted by **adolwyn** March 25, 2026</sup>
### What Went Wrong?

Some books in Grimmory are being imported with the title set to “A Novel” instead of the book’s actual title.

This appears to be related to a previous Booklore issue where "A Novel" was incorrectly being treated as the main title, although my case seems to involve a slightly different metadata pattern and may not be covered by the earlier fix.

In my case, the affected books were managed in Calibre. I changed the visible title there so that the actual title no longer included "A Novel", and then exported the books to disk. However, when those exported books are imported into Grimmory, some of them still appear with the title “A Novel”.

After inspecting the exported metadata, I found that the .opf metadata still contains a subtitle-related title entry like this:
`<dc:title id="subtitle">A Novel</dc:title>`

That makes me suspect Grimmory is still treating this subtitle-related dc:title entry as the primary title in some cases.

From what I remember of the older Booklore issue, the previous case involved a structure more like:
	•	primary title in something like dc:title id="t1"
	•	subtitle in something like dc:title id="t2"

My affected books seem to be different:
	•	the real main title exists separately
	•	the subtitle specifically appears as dc:title id="subtitle"

So this may be a related issue, but with a different OPF metadata pattern that is not being handled correctly.

I also noticed there are tests in the Grimmory / former Booklore codebase showing that subtitle is already treated as a distinct metadata concept in other parts of the application, which is why I suspect the issue is specifically in metadata import/parsing rather than the overall data model.

Examples I found:
	•	frontend/src/app/features/magic-shelf/service/book-rule-evaluator.service.spec.ts
	•	uses subtitle: 'A Novel'
	•	frontend/src/app/features/magic-shelf/service/book-rule-evaluator-metadata-presence.spec.ts
	•	uses subtitle: 'A Novel'
	•	booklore-api/src/test/java/org/booklore/service/metadata/writer/BookLoreSchemaTest.java
	•	includes a test where schema.setSubtitle("A Novel")

Those do not prove where the bug is, but they do suggest subtitle is already expected to exist as its own field, which makes me think the problem is likely in EPUB metadata parsing/import where dc:title id="subtitle" may still be getting promoted to the main title.

### How Can We Reproduce This?

	1.	Take an EPUB (or other imported ebook) whose OPF metadata contains a subtitle entry like:
`<dc:title id="subtitle">A Novel</dc:title>`

	2.	Ensure the actual main title is something else, for example Example Title.
	3.	Export the book from Calibre to disk.
	4.	Import that exported book into Grimmory.
	5.	Grimmory displays the title as “A Novel” instead of the actual book title.

In my case, I did not need to do anything unusual or perform any special metadata-editing steps beyond normal Calibre metadata editing and export. I simply imported the book into Grimmory and it came in with the incorrect title.

### What Should Have Happened?

Grimmory should import and display the actual primary title of the book.

If the OPF metadata contains a subtitle entry such as:
`<dc:title id="subtitle">A Novel</dc:title>`

that value should either:
	•	remain associated with subtitle metadata only, or
	•	be ignored for purposes of determining the main title

It should not replace the primary title.

### Screenshots or Error Messages (Optional)

I do not currently have an error message from Grimmory for this issue.

I will

Image
Image
Image

 also provide screenshots showing that:
	•	the title is correct in Calibre before export
	•	the OPF still contains the subtitle tag
	•	Grimmory imports the book with the title shown as "A Novel"

### Any Ideas on How to Fix This? (Optional)

My guess is that the import/parsing logic may still be treating some dc:title entries as possible main titles without correctly distinguishing subtitle metadata from the actual primary title.

A possible fix might be to:
	•	prioritize the true primary title field when multiple dc:title values exist
	•	ignore dc:title entries marked as subtitle metadata when determining the main display title
	•	compare this case against the earlier Booklore fix to see whether the old fix handled only patterns like t1 / t2, but not subtitle

It may also help to test import behaviour specifically against OPF files that use:
	•	dc:title id="title"
	•	dc:title id="subtitle"
	•	multiple dc:title entries with different IDs/refinements

to confirm which field Grimmory currently treats as canonical.

### Your Setup

- Grimmory version: 2.3.0
- Install method: Docker (using Orbstack on a mac)
- OS: Mac Sequoia 15.6
- Browser: Chrome
- Import source: books exported from Calibre to disk
- Storage Type: spinning external hard drive (not SSD)

### Before Submitting

- [x] I've searched existing issues and confirmed this bug hasn't been reported yet</div>

### PR DESCRIPTION
**Pull request by imnotjames**
## Description

Epub title parsing currently relies heavily on the epub3 `refines` meta to understand what is the primary title of a book.  However, in cases where this is not available, the epub spec notes:

> If no means of determining title types is provided, or understood, Reading Systems must treat the first title element in document order as the main title. This specification does not define how additional title elements should be processed in such situations.

This change updates our epub parsing to default the first title by-order as the fallback main title, per spec.

## Linked Issue

fixes #1636 

## Changes

* Added handling for case where multiple titles exist but do not have refines.
* Updated tests to check various edge cases

## Manual Testing Steps

1. Uploaded books matching each test case (multiple titles, no refines, multiple titles with refines in different orders)
2. Validated values are interpreted as expected

## AI Disclosure

None.

## Checklist

- [x] This PR links and implements an accepted issue.
- [x] This PR is a single focused change.
- [x] There are new or updated tests validating this change.
- [x] I ran `just ui check` and `just api check`.
- [x] I have added screenshots if there were any UI changes.
- [x] I have disclosed any AI usage as per the organization AI Policy above.
- [x] I understand all of my submitted changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved EPUB title/subtitle detection to prevent incorrect overwrites and misassignment when type information is missing or titles are duplicated or blank; ensures sensible fallback behavior.
* **Tests**
  * Added unit tests validating title/subtitle selection order, metadata-refinement precedence, and fallback behavior for missing or blank metadata.

### CODE REVIEW

**Comment by imnotjames** [comment]
I am iffy about the second title found being the subtitle and am going to sit on that for a bit and think about it.  I can see cases where multiple title languages are stored and this will lead to improper subtitles.

File: backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java [hunk 1]
```diff
@@ -186,6 +186,48 @@ void distinguishesMainTitleFromSubtitleViaRefines() throws IOException {
...
+                    <dc:title id="t4">The Subtitle</dc:title>
+                    <dc:title>Even more titles</dc:title>
+                    <meta refines="#t2" property="title-type">main</meta>
+                    <meta refines="#t4" property="title-type">subtitle</meta>
+                    """);
+            BookMetadata metadata = extractor.extractMetadata(createEpub(opf));
+
+            assertThat(metadata.getTitle()).isEqualTo("Main Title");
+            assertThat(metadata.getSubtitle()).isEqualTo("The Subtitle");
+        }
```
[the remark this answers is not in the document]
Comment (imnotjames) [reply backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java:229]: I think this is an incorrect test case, given the epub specification doesn't say anything about blank title elements being ignored.

[the remark this answers is not in the document]
Comment (imnotjames) [reply backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java:229]: Maybe

> This specification imposes no additional restrictions or requirements on titles except that they must be at least one character in length. 

But that just means that the epub manifest is invalid.  That does not mean that we do something off spec, do we?

I'll investigate the existing behaviors.

[the remark this answers is not in the document]
Comment (imnotjames) [reply backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java:229]: I added this test but validated it uses the expected behavior.

File: backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java [hunk 2]
```diff
@@ -186,6 +186,60 @@ void distinguishesMainTitleFromSubtitleViaRefines() throws IOException {
...
+        void blankTitlesFallbackToFilename() throws IOException {
+            String opf = wrapOpf("""
+                    <dc:title> </dc:title>
+                    <dc:title>The Subtitle</dc:title>
+                    """);
+            BookMetadata metadata = extractor.extractMetadata(createEpub(opf));
+
+            assertThat(metadata.getTitle()).isEqualTo("test");
+            assertThat(metadata.getSubtitle()).isEqualTo("The Subtitle");
+        }
```
[the remark this answers is not in the document]
Comment (imnotjames) [reply backend/src/test/java/org/booklore/service/metadata/extractor/EpubMetadataExtractorTest.java:241]: I am ignoring you for this thanks bye

**Review by zachyale — APPROVED** [review]
